### PR TITLE
Add horizontal and vertical compaction tasks for event_based compaction framework

### LINKED ISSUE #3365: Add HorizontalCompactionTask and VerticalCompactionTask for new compaction framework
## Enhancement
add two type compaction tasks, HorizontalCompactionTask and VerticalCompactionTask

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this：
- [ ] bug
- [X] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #3365 

## Problem Summary(Required) ：

The original compaction framework polled the candidate tablets one by one. The performance is low and overhead is high.
Now I change the mechanism of picking compaction candidates from polling to triggering by the event(#3311).
Considering StarRocks split the columns of a table into groups,  it can improve the performance(https://github.com/StarRocks/starrocks/pull/2189) of compaction. This pull request adds the horizontal and vertical compaction to the event_based compaction framework.

### CODE REVIEW

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/storage/horizontal_compaction_task.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        RETURN_IF_ERROR(output_rs_writer->add_chunk(*chunk));
+        output_rows += chunk->num_rows();
+        _task_info.output_num_rows = output_rows;
+        _task_info.filtered_rows = reader.stats().rows_del_filtered;
+        _task_info.merged_rows = reader.merged_rows();
+    }
+    TRACE("[Compaction] data compacted");
+
+    if (should_stop()) {
+        LOG(INFO) << "horizontal compaction task_id:" << _task_info.task_id << " is stopped.";
```
Comment (chaoyli) [inline be/src/storage/horizontal_compaction_task.cpp:144]: Should add tablet_id info in log.

File: be/src/storage/vertical_compaction_task.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+    auto mask_buffer =
+            std::make_unique<vectorized::RowSourceMaskBuffer>(_tablet->tablet_id(), _tablet->data_dir()->path());
+    auto source_masks = std::make_unique<std::vector<vectorized::RowSourceMask>>();
+    TRACE("[Compaction] compaction prepare finished, max_rows_per_segment:$0, column groups "
+          "size:$1",
+          max_rows_per_segment, column_groups.size());
+
+    for (size_t i = 0; i < column_groups.size(); ++i) {
+        if (should_stop()) {
```
Comment (chaoyli) [inline be/src/storage/vertical_compaction_task.cpp:63]: Should add tablet_id in the log.

File: be/src/storage/horizontal_compaction_task.cpp [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    vectorized::TabletReader reader(std::static_pointer_cast<Tablet>(_tablet->shared_from_this()),
+                                    output_rs_writer->version(), schema);
+    vectorized::TabletReaderParams reader_params;
+    reader_params.reader_type = ReaderType::READER_LEVEL_COMPACTION;
+    reader_params.profile = _runtime_profile.create_child("merge_rowsets");
+
+    int32_t chunk_size = CompactionUtils::get_read_chunk_size(
+            config::compaction_memory_limit_per_worker, config::vector_chunk_size, _task_info.input_rows_num,
+            _task_info.input_rowsets_size, _task_info.segment_iterator_num);
+    VLOG(1) << "compaction task_id:" << _task_info.task_id << ", tablet=" << _tablet->tablet_id()
```
Comment (chaoyli) [inline be/src/storage/horizontal_compaction_task.cpp:59]: What does the task_id is used for? What message do you want to give the user?

Comment (ABingHuang) [reply be/src/storage/horizontal_compaction_task.cpp:59]: the task_id is unique for each compaction task. It has the following benefits:
1. useful to locate compaction task in log for debug
2. can now how many compaction tasks there is
3. can be used to cancel the specific compaction task in the future.

File: be/src/storage/vertical_compaction_task.cpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+                                                       std::vector<vectorized::RowSourceMask>* source_masks) {
+    size_t output_rows = 0;
+    auto chunk = vectorized::ChunkHelper::new_chunk(schema, chunk_size);
+    auto char_field_indexes = vectorized::ChunkHelper::get_char_field_indexes(schema);
+
+    Status status = Status::OK();
+    size_t column_group_del_filtered_rows = 0;
+    size_t column_group_merged_rows = 0;
+    while (!should_stop()) {
+#ifndef BE_TEST
```
Comment (meegoo) [inline be/src/storage/vertical_compaction_task.cpp:190]: using UNLIKELY in this loop might help

Comment (ABingHuang) [reply be/src/storage/vertical_compaction_task.cpp:190]: ok

Comment (ABingHuang) [reply be/src/storage/vertical_compaction_task.cpp:190]: I think it is LIKELY here

**Review by chaoyli — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Review by decster — APPROVED** [review]

**Pull request merged**